src/types/CodepointWidthDetector.cpp

```diff
@@ -4,398 +4,1028 @@
+
+        auto w = ucdToCharacterWidth(val);
+        if (w == 3)
+        {
+            w = _ambiguousWidth;
+        }
+
+        if (state != 0 && w != 0)
+        {
+            state = 0;
+            break;
+        }
+
+        width += w;
+        state = 1;
+        clusterEnd = clusterEndNext;
+
+        if (clusterEnd >= end)
         {
-            width = _checkFallbackViaCache(codepoint, glyph);
+            break;
         }
     }
 
-    return width;
+    s.beg = clusterBeg;
+    s.len = static_cast<int>(clusterEnd - clusterBeg);
+    s.width = width;
+    s._state = state;
+    return clusterEnd < end;
 }
 
-// Call the function specified via SetFallbackMethod() to turn CodepointWidth::Ambiguous into Narrow/Wide.
-// Caches the results in _fallbackCache. This is _lookupGlyphWidth's even-slower-path.
-uint8_t CodepointWidthDetector::_checkFallbackViaCache(const char32_t codepoint, const std::wstring_view& glyph) noexcept
+__declspec(noinline) bool CodepointWidthDetector::_graphemePrevWcswidth(GraphemeState& s, const wchar_t* beg, const wchar_t* clusterEnd) noexcept
+{
+    if (_mode != TextMeasurementMode::Wcswidth)
+    {
+        return _graphemePrevConsole(s, beg, clusterEnd);
+    }
+
+    auto clusterBeg = clusterEnd;
+    auto width = s.width;
+    int state = 0;
+
+    // Insert a `.len = 0` result if we previously returned false but had a complete cluster (`.width > 0`).
+    // This will indicate to the caller that the previous cluster was complete, just like how the grapheme algorithm works.
+    if (s._state == 0)
+    {
+        width = 0;
+
+        for (;;)
+        {
+            char32_t cp;
+            clusterBeg = utf16PrevOrFFFD(clusterBeg, beg, cp);
+            const auto val = ucdLookup(cp);
+
+            auto w = ucdToCharacterWidth(val);
+            if (w == 3)
+            {
+                w = _ambiguousWidth;
+            }
+
+            width += w;
+
+            const auto hasWidth = width != 0;
+            const auto atEnd = clusterBeg <= beg;
+
+            if (hasWidth || atEnd)
+            {
+                state = hasWidth && atEnd;
+                break;
+            }
+        }
+    }
+
+    s.beg = clusterBeg;
+    s.len = static_cast<int>(clusterEnd - clusterBeg);
+    s.width = width;
+    s._state = state;
+    return clusterBeg > beg;
+}
+
+bool CodepointWidthDetector::_graphemeNextConsole(GraphemeState& s, const wchar_t* end, const wchar_t* clusterBeg) noexcept
+{
+    auto clusterEnd = clusterBeg;
+    auto width = s.width;
+    int state = 0;
+
+    // Insert a `.len = 0` result if we previously returned false but had a complete cluster (`.width > 0`).
+    // This will indicate to the caller that the previous cluster was complete, just like how the grapheme algorithm works.
+    if (s._state == 0)
+    {
+        char32_t cp;
+        clusterEnd = utf16NextOrFFFD(clusterEnd, end, cp);
+
+        const auto val = ucdLookup(cp);
+        width = ucdToCharacterWidth(val);

```

Comment:
i left a comment on the table generator above about the overrides, but now I see that the old table is totally gone _even in the `Console` measurement case_. This will break those overridden ranges.

Reply:
(Will close this in favor of the other thread to continue the discussion there.)